wireguard/metrics.go

```diff
@@ -0,0 +1,284 @@
+
+	lastCollectionTime time.Time
+	rateLimitInterval  time.Duration
+}
+
+func (collector *Metrics) Describe(d chan<- *prometheus.Desc) {
+	for _, dev := range collector.getDevices() {
+		collector.descsByDevice(d, dev)
+	}
+}
+
+func (collector *Metrics) descsByDevice(d chan<- *prometheus.Desc, device *wgtypes.Device){
+	if device == nil {
+		collector.logCtx.Error("BUG: called descsByDevice with nil device")
+		return
+	}
+
+	labels := collector.defaultLabelValues("pub", deviceMetaLabelValues(device))
+	for fqName, help := range map[string]string{
+		wireguardMetaFQName: wireguardMetaHelpText,
+	} {
+		d <- prometheus.NewDesc(fqName, help, nil, labels)
+	}
+	for _, peer := range device.Peers {
+		collector.descByPeer(d, &peer)
+	}
+}
+
+func (collector *Metrics) descByPeer(d chan<- *prometheus.Desc, peer *wgtypes.Peer) {
+	if peer == nil {
+		collector.logCtx.Error("BUG: called descByPeer with nil peer")
+		return
+	}
+
+	labels := collector.defaultLabelValues("pub", peerServiceLabelValues(peer))
+	for fqName, help := range map[string]string{
+		wireguardBytesRcvdFQName: wireguardBytesRcvdHelpText,
+		wireguardBytesSentFQName: wireguardBytesSentHelpText,
+		wireguardLatestHandshakeIntervalFQName: wireguardLatestHandshakeIntervalHelpText,
+	} {
+		d <- prometheus.NewDesc(fqName, help, nil, labels)
+	}
+}
+
+func (collector *Metrics) Collect(m chan<- prometheus.Metric) {
+	collector.refreshStats(m)
+}
+
+func MustNewWireguardMetrics() *Metrics {
+	wg, err := NewWireguardMetrics()
+	if err != nil {
+		logrus.Panic(err)
+	}
+	return wg
+}
+
+func NewWireguardMetrics() (*Metrics, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		logrus.WithError(err).Error("cannot register wireguard metrics stats")
+		return nil, err
+	}
+	return NewWireguardMetricsWithShims(hostname, netlinkshim.NewRealWireguard, defaultCollectionRatelimit), nil
+}
+
+func NewWireguardMetricsWithShims(hostname string, newWireguardClient func() (netlinkshim.Wireguard, error), rateLimitInterval time.Duration) *Metrics {
+	logrus.WithField("hostname", hostname).Debug("created wireguard collector for host")
+	return &Metrics{
+		hostname:           hostname,
+		newWireguardClient: newWireguardClient,
+		logCtx: logrus.WithFields(logrus.Fields{
+			"prometheus_collector": "wireguard",
+		}),
+
+		peerRx: map[wgtypes.Key]int64{},
+		peerTx: map[wgtypes.Key]int64{},
+
+		rateLimitInterval: rateLimitInterval,
+	}
+}
+
+func (collector *Metrics) getDevices() []*wgtypes.Device {
+	wgClient, err := collector.newWireguardClient()
+	if err != nil {
+		collector.logCtx.WithError(err).Error("error initializing wireguard client devices")
+		return nil
+	}
+
+	devices, err := wgClient.Devices()
+	if err != nil {
+		collector.logCtx.WithError(err).Error("error listing wireguard devices")
+		return nil
+	}
+	return devices
+}
+
+func (collector *Metrics) refreshStats(m chan<- prometheus.Metric) {
+	if ct := time.Since(collector.lastCollectionTime); ct < collector.rateLimitInterval {
```

Comment:
What does this actually do to prometheus?  Will it show as a gap in the graph?  I was imagining caching the previous value but I  guess that'd be fiddly

Reply:
as far as my testing went it does not show as a gap since it's only 5s.. if you select a crazy query range step like 1s _maybe_ you can see a teeny gap but most tools like grafana smooth that out anyway. 

if we bring it further down to 1s (which is the smallest resolution in prom anyway), nothing will be affected then and will just prevent DDoS-type calls to dataplane